lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,241 @@
+        post[:transaction][:invoice] = options[:transaction][:invoice]
+        post[:transaction][:notes] = options[:transaction][:notes] if options[:transaction][:notes].present?
+        if options[:transaction][:purchase_card].present?
+          post[:transaction][:purchaseCard] = {}
+          post[:transaction][:purchaseCard][:customerReference] = options[:transaction][:purchase_card][:customer_reference]
+          post[:transaction][:purchaseCard][:destinationPostalCode] = options[:transaction][:purchase_card][:destination_postal_code]
+          post[:transaction][:purchaseCard][:productDescriptors] = options[:transaction][:purchase_card][:product_descriptors]
+        end
+        post[:overrideBusDate] = options[:override_bus_date] if options[:override_bus_date].present?
+        if post[:amex].present?
+          post[:amex] = {}
+          post[:amex][:propertyCode] = options[:amex][:property_code]
+        end
+        if post[:auto].present?
+          post[:auto] = {}
+          post[:auto][:estimatedDays] = options[:auto][:estimated_days]
+        end
+        if post[:cardOnFile].present?
+          post[:cardOnFile] = {}
+          post[:cardOnFile][:indicator] = options[:card_on_file][:indicator]
+          post[:cardOnFile][:usageIndicator] = options[:card_on_file][:usage_indicator]
+          post[:cardOnFile][:scheduledIndicator] = options[:card_on_file][:scheduled_indicator]
+          post[:cardOnFile][:indicator] = options[:card_on_file][:transaction_id]
+        end
+        if post[:hotel].present?
+          post[:hotel] = {}
+          post[:hotel][:estimatedDays] = options[:hotel][:estimated_days]
+        end
+        if post[:pin].present?
+          post[:pin] = {}
+          post[:pin][:block] = options[:pin][:block]
+          post[:pin][:ksn] = options[:pin][:ksn]
+        end
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, options)
+        post[:card] = {}
+        if options[:card][:token].present?
```

Comment:
This method should be refactored to use the payment_method object being passed in to the transaction method.

Reply:
Done